public/js/monitor.js

```diff
@@ -413,47 +408,111 @@ function addWaitlistObservers() {
+  } else {
+    vpnBanner.querySelector(".client-ip").remove();
+  }
 
-  if (vpnBannerDismissedCookie) {
-    return;
+  if (locationData?.shortLocation) {
+    vpnBanner.querySelector(".short-location output").textContent = locationData.shortLocation;
+  } else {
+    vpnBanner.querySelector(".short-location").remove();
   }
 
-  // Init: Show banner, set close button listener
-  const vpnPromoCloseButton = document.getElementById("vpnPromoCloseButton");
-
-  const vpnPromoFunctions = {
-    hide: function() {
-      vpnPromoFunctions.setCookie();
-      vpnPromoBanner.classList.add("closed");
-      document.body.classList.remove("vpn-banner-visible");
-    },
-    init: function() {
-      vpnPromoCloseButton.addEventListener("click", vpnPromoFunctions.hide);
-      vpnPromoFunctions.show();
-    },
-    setCookie: function() {
-      const date = new Date();
-      date.setTime(date.getTime() + 30*24*60*60*1000);
-      document.cookie = "vpnBannerDismissed=true; expires=" + date.toUTCString();
-    },
-    show: function() {
-      vpnPromoBanner.classList.remove("closed");
-      document.body.classList.add("vpn-banner-visible");
-    },
-  };
+  if (locationData?.fullLocation) {
+    vpnBanner.querySelector(".full-location output").textContent = locationData.fullLocation;
+  } else {
+    vpnBanner.querySelector(".full-location").remove();
+  }
+
+  vpnBanner.cta = vpnBanner.querySelector("a.vpn-banner-cta");
+  vpnBanner.cta.setAttribute("href", vpnBanner.cta.getAttribute("href") + getPageAttribution());
+  vpnBanner.setAttribute("data-protected", Boolean(protectionData?.isProtected));
+  vpnBanner.addEventListener("click", handleClick);
+
+  if (cache && protectionData) cache.put(protectionDataReq, new Response(JSON.stringify(protectionData)));
+
+  async function initCache() {
+    const cacheAvailable = "caches" in self;
+    let cache;
+
+    if (cacheAvailable) cache = await caches.open("vpn-banner").catch(e => null);
+
+    return cache;
+  }
 
-  vpnPromoFunctions.init();
+  async function getCacheData(req) {
+    if (!cache) return null;
+
+    const json = await cache.match(req)
+      .then(res => res.json())
+      .catch(e => console.warn("Could not get cached response.", e.message));
+
+    return json;
+  }
+
+  async function fetchData(req) {
+    const abortController = new AbortController();
+    const timer = setTimeout(() => abortController.abort(), 4000); // abort a delayed response after 4s
+    const json = await fetch(req, { signal: abortController.signal })
+      .then(res => {
+        clearTimeout(timer);
+        if (!res.ok) throw new Error(`Bad response (${res.status})`);
+        return res.json();
+      })
+      .catch(e => console.warn("Error fetching protection data.", e));
+
+    return json;
+  }
+
+  function getPageAttribution() {
+    let page = location.pathname;
+
+    if (page.startsWith("/")) page = page.slice(1);
+
+    if (page === "") page = "home";
+
+    return `&utm_content=${page}`;
```

Comment:
Does this account for query strings that don't have a "?" or could we end up w/ URLs like `https://monitor.firefox.com&utm_content=home`?

Reply:
The link is hard-coded: 
`https://vpn.mozilla.org/?utm_source=firefox-monitor&utm_medium=banner&utm_campaign=ip`

The `utm_content` key is the only dynamic value.  If you agree, it might be more intuitive to add the entire query string in js, instead of splitting the static/dynamic parts between html/js.